File: cpp/adapter.cpp
#include <iostream>
#include <string>
using namespace std;

class InternalClass {
public:
    virtual string fetch() {   // <-- virtual 추가
        return "get user info";
    }
    virtual ~InternalClass() = default; // 소멸자도 virtual이면 좋음
};

class ExternalClass {
public:
    string search() {
        return "get user info";
    }
};

class Adapter : public InternalClass {
private:
    ExternalClass* external;
public:
    Adapter(ExternalClass* external) : external(external) {}

    string fetch() override {   // 이제 제대로 오버라이드 됨
        return external->search();
    }
};

int main() {
    InternalClass internal;
    cout << internal.fetch() << endl; // get user info

    ExternalClass external;
    Adapter adapter(&external);
    cout << adapter.fetch() << endl; // get user info
}

File: kotlin/adapter.kt
interface Internal {
    fun fetch(): String
}

class InternalClass : Internal {
    override fun fetch(): String {
        return "get user info"
    }
}

class ExternalClass {
    fun search(): String {
        return "get user info"
    }
}

class Adapter(private val external: ExternalClass) : Internal {
    override fun fetch(): String {
        return external.search()
    }
}

fun main() {
    val internal: Internal = InternalClass()
    println(internal.fetch()) // get user info

    val external = ExternalClass()
    val adapter: Internal = Adapter(external)
    println(adapter.fetch()) // get user info
}
